### PR TITLE
Proxy implementation refactor to include queries against an observer

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added possibility to query an observer rather than a proxy
- implemented a new way to check the finality status of a shard
- major code refactor in proxy implementation

### CODE REVIEW

File: blockchain/endpointProviders/nodeEndpointProvider.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import "fmt"
+
+const (
+	nodeGetNodeStatusEndpoint      = "node/status"
+	nodeRawBlockByHashEndpoint     = "internal/raw/block/by-hash/%s"
+	nodeRawBlockByNonceEndpoint    = "internal/raw/block/by-nonce/%d"
+	nodeRawMiniBlockByHashEndpoint = "internal/raw/miniblock/by-hash/%s/epoch/%d"
+)
+
+// proxyEndpointProvider is suitable to work with an Elrond node
```
Comment (dragos-rebegea) [inline blockchain/endpointProviders/nodeEndpointProvider.go:12]: fix comment

File: blockchain/interface.go [hunk 2]
```diff
@@ -22,3 +22,33 @@ type httpClientWrapper interface {
 	PostHTTP(ctx context.Context, endpoint string, data []byte) ([]byte, int, error)
 	IsInterfaceNil() bool
 }
+
+// EndpointProvider is able to return endpoint routes strings
+type EndpointProvider interface {
```
Comment (dragos-rebegea) [inline blockchain/interface.go:27]: maybe remove the Endpoint keyword from methods name. You will anyway call it as endpointProvider.GetVmValuesEndpoint, so it should already makes sense that you are actually geting an endpoint.

Comment (iulianpascalau) [reply blockchain/interface.go:27]: removed the `Endpoint` suffix from all methods

File: blockchain/endpointProviders/nodeEndpointProvider.go (hunk 1 above)
Comment (iulianpascalau) [reply blockchain/endpointProviders/nodeEndpointProvider.go:12]: done

File: blockchain/factory/factory.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+// CreateFinalityProvider creates a new instance of FinalityProvider
+func CreateFinalityProvider(entityType core.RestAPIEntityType, proxy proxy, checkFinality bool) (FinalityProvider, error) {
+	if !checkFinality {
+		return finalityProvider.NewDisabledFinalityProvider(), nil
+	}
+
+	switch entityType {
+	case core.ObserverNode:
+		return finalityProvider.NewNodeFinalityProvider(proxy)
```
Comment (raduchis) [inline blockchain/factory/factory.go:31]: shouldn't the entityType and proxyEndpointProvider type be of the same value? could it be taken from the proxy?
can you have a core.Proxy endpointProvider in the proxy with a nodeFinalityProvider?

File: blockchain/endpointProviders/nodeEndpointProvider.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// NewNodeEndpointProvider returns a new instance of a nodeEndpointProvider
+func NewNodeEndpointProvider() *nodeEndpointProvider {
+	return &nodeEndpointProvider{}
+}
+
+// GetNodeStatus returns the node status endpoint
+func (node *nodeEndpointProvider) GetNodeStatus(_ uint32) string {
+	return nodeGetNodeStatusEndpoint
```
[the remark this answers is not in the document]
Comment (raduchis) [reply blockchain/endpointProviders/nodeEndpointProvider.go:24]: talked about it and it should be fine

[the remark this answers is not in the document]
Comment (iulianpascalau) [reply blockchain/endpointProviders/nodeEndpointProvider.go:24]: as stated, as the check is pretty hard to be implemented when the request is being done on a proxy without a very ugly if in the proxy implementation. Will defeat the purpose of having decoupled components. Will think of a solution as I do not have a great one now. Added TODO

File: blockchain/factory/factory.go (hunk 3 above)
Comment (iulianpascalau) [reply blockchain/factory/factory.go:31]: done

**Review by raduchis — APPROVED** [review]

**Review by dragos-rebegea — APPROVED** [review]

**Pull request merged**